File: problem_2/problem_2_even_fibonacci_numbers.exs
defmodule Math do

  # Still working on a proper way to get the stream to stop if the sum of the
  # last two numbers in the sequence is >= 4000000.
  def create_sequence do
    Stream.unfold({1, 2}, fn {x, y} -> {x, {y, x + y}} end)
    |> Enum.take(32)
  end

  def even_numbers_only do
    Enum.filter(create_sequence, &even?/1)
  end

  def even?(x) do
    rem(x, 2) == 0
  end

  def perform do
    even_numbers_only
    |> Enum.sum
  end

end

File: problem_2/problem_2_even_fibonacci_numbers.rb
class SumFibonacci

  def sum_even_numbers
    self.send(:even_numbers_only).reduce(:+)
  end

  private

    def create_sequence
      sequence = [1, 2]

      while (sequence[-1] + sequence[-2]) < 4000000
        calculate_next_term(sequence)
      end

      sequence
    end

    def calculate_next_term(sequence)
      sequence << (sequence[-1] + sequence[-2])
    end

    def even_numbers_only
      self.send(:create_sequence).keep_if { |n| n % 2 == 0 }
    end

end
